pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java

```diff
@@ -0,0 +1,98 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.calcite.sql.fun;
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.calcite.sql.SqlFunction;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.validate.SqlNameMatchers;
+import org.apache.calcite.util.Util;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+
+
+/**
+ * {@link PinotOperatorTable} defines the {@link SqlOperator} overrides on top of the {@link SqlStdOperatorTable}.
+ *
+ * <p>The main purpose of this Pinot specific SQL operator table is to
+ * <ul>
+ *   <li>Ensure that any specific SQL validation rules can apply with Pinot override entirely over Calcite's.</li>
+ *   <li>Ability to create customer operators that are not function and cannot use
+ *     {@link org.apache.calcite.prepare.Prepare.CatalogReader} to override</li>
+ *   <li>Still maintain minimum customization and benefit from Calcite's original operator table setting.</li>
+ * </ul>
+ */
+public class PinotOperatorTable extends SqlStdOperatorTable {
+
+  private static @MonotonicNonNull PinotOperatorTable _instance;
+
+  public static final SqlFunction COALESCE = new PinotSqlCoalesceFunction();
+
+  public static synchronized PinotOperatorTable instance() {
+    if (_instance == null) {
+      // Creates and initializes the standard operator table.
+      // Uses two-phase construction, because we can't initialize the
+      // table until the constructor of the sub-class has completed.
+      _instance = new PinotOperatorTable();
+      _instance.initNoDuplicate();
+    }
+    return _instance;
+  }
+
+  /**
+   * Initialize without duplicate, e.g. when 2 duplicate operator is linked with the same op
+   * {@link org.apache.calcite.sql.SqlKind} it causes problem.
+   */
+  public final void initNoDuplicate() {
+    // Use reflection to register the expressions stored in public fields.
+    for (Field field : getClass().getFields()) {
```

Comment:
😬 instead of using reflection can we just have them all in an enum?
```
enum Functions {
  COALESCE(new PinotSqlCoalesceFunction());

  final SqlOperator _op;

  Functions(SqlOperator op) {...}

  void register(PinotOperatorTable table) {...}
}
```

and then just call:
```
for (Functions fun : Functions.values()) {
  fun.register(this);
}
```

Also is there any reason we differentiate between SqlFunction and SqlOperator? it looks like the code handling them is the same.

Reply:
yeah SqlStdOperatorTable is a reflection-based, i was just copying to code out and modify it. 
this is one-time lazy static initialization, so I will leave this in place and add a TODO for perf improvement

Reply:
sounds good 👍 it was more for clarity (reflection is hard to debug) than for performance but I suspect this is pretty low chance of failure.

Reply:
Should we just use SqlFunction for now since that's the only thing we have in fields? 
Would be good to add a comment to caution on adding fields since this will be used in this func and this needs to be no-dup and it is case-insensitive.

Reply:
no we have operators such as LIKE and REGEXP_LIKE